core-rust/state-manager/src/store/parent_map.rs

```diff
@@ -0,0 +1,315 @@
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use crate::store::traits::*;
+use crate::{ChangeAction, SubstateChange};
+
+use radix_engine::types::*;
+
+/// A resolver of [`SubstateNodeMetadata`] for [`SubstateChange`]s.
+pub struct NodeMetadataResolver {
+    /// A map of created Nodes to their parent Substates, extracted from the [`SubstateChange`]s.
+    changed_parent_by_node: NonIterMap<NodeId, SubstateReference>,
+    /// A map of existing Nodes to their root Nodes (i.e. `SubstateNodeMetadata#root`), loaded from
+    /// a [`SubstateNodeMetadataStore`] for each Node that does not have a parent in the
+    /// [`changed_parent_by_node`] map.
+    existing_root_by_node: NonIterMap<NodeId, SubstateReference>,
+}
+
+impl NodeMetadataResolver {
+    /// Resolves the [`SubstateNodeMetadata`] for all newly-created Nodes found in the [`Create`]d
+    /// and [`Update`]d substates. (within the given [`SubstateChange`]s).
+    /// The resolution considers the trees (or rather: potentially-incomplete bottom-up tree
+    /// fragments) constructed within the given [`SubstateChange`]s, and the top fragments of these
+    /// trees already-existing in the given [`SubstateNodeMetadataStore`].
+    /// API note: the results are grouped by [`SubstateNodeMetadata`] (since it may be shared by
+    /// many [`NodeId`]s).
+    pub fn batch_resolve<S: SubstateNodeMetadataStore>(
+        metadata_store: &S,
+        substate_changes: &[SubstateChange],
+    ) -> impl Iterator<Item = (Vec<NodeId>, SubstateNodeMetadata)> {
+        // Index the new Nodes by their parent Substates (using `IndexedScryptoValue`):
+        let changed_nodes_by_parent = substate_changes
+            .iter()
+            .filter_map(|substate_change| match &substate_change.action {
+                ChangeAction::Create(bytes) | ChangeAction::Update(bytes) => {
+                    let owned_node_ids = IndexedScryptoValue::from_slice(bytes).unwrap().unpack().1;
+                    if owned_node_ids.is_empty() {
+                        return None;
+                    }
+                    let parent_reference = SubstateReference(
+                        substate_change.node_id,
+                        substate_change.partition_number,
+                        substate_change.substate_key.clone(),
+                    );
+                    Some((parent_reference, owned_node_ids))
+                }
+                ChangeAction::Delete => None,
+            })
+            .collect::<IndexMap<_, _>>();
```

Comment:
I'm slightly worried about edge case behaviour here - EG conceivably one substate could have 1000 children and updates some irrelevant data - but we're still querying our DB for all 1000 children each update here.

It may be more efficient to filter to only nodes that have been created here... Which I thought we could easily get from the receipt, but looks like we don't have that information available :'(.

Although actually, I believe we could also resolve the "previous" values for updates in `compute_substate_changes`, right? (and in fact, maybe we should even persist this as part of `ChangeAction`? This would be helpful for Gateway for sure)

If we had that, we could compare "before" and "after", and only need to resolve/add newly created nodes, which would limit the number of lookups we need to do here quite a lot I imagine.

Reply:
> but we're still querying our DB for all 1000 children each update here.

I don't see this.
I definitely query the DB about the _parent_ of these children (i.e. the parent == the substate that was created/updated here), in a 1 batch read, to discover the root of that parent.

The inefficiency I see though:
We will create 1000 Metadata rows, and we will upsert them on commit, which will be completely needless since they already exist :(

I agree that this could be mitigated by adding `previous` to `ChangeAction::Update()`.
It's definitely worthwhile - should we treat it as a performance optimization that happens in next PR?

Reply:
Ahh right, I didn't look far enough down to work out what the inefficiency actually was, but knew we needed to handle these 1000 things somehow. Happy to do it in the next PR.